network/p2p/router.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/message"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/version"
+)
+
+var (
+	ErrExistingAppProtocol = errors.New("existing app protocol")
+	ErrUnrequestedResponse = errors.New("unrequested response")
+
+	_ common.AppHandler    = (*Router)(nil)
+	_ validators.Connector = (*Router)(nil)
+)
+
+// Router routes incoming application messages to the corresponding registered
+// app handler. App messages must be made using the registered handler's
+// corresponding Client.
+type Router struct {
+	log    logging.Logger
+	sender common.AppSender
+
+	lock                         sync.RWMutex
+	handlers                     map[uint64]*responder
+	pendingAppRequests           map[uint32]AppResponseCallback
+	pendingCrossChainAppRequests map[uint32]CrossChainAppResponseCallback
+	requestID                    uint32
+	peers                        set.SampleableSet[ids.NodeID]
```

Comment:
I think we could use an interface for this type that has eg, Connected, Disconnected (or Add/Remove) and SampleOne.
This would allow us to eg, combine this type with the peer tracker.
It may also simplify locking to assume this interface handles its own locking.